Suppress backlog clutter

glubsy opened the pull request:
* Overwrite Streamlink's output after every call that doesn't return an exit code of 0.
This should preserve Streamlink's output when a live stream has been detected.
For every other call where a live stream is not active, we safely overwrite the repeating output and keep only the last two lines.
To achieve this we use special non-printable characters to tell the terminal to go back two columns up, clear the line to the beginning, write the date, and clear the rest of the line.
The special characters might be different from one terminal emulator to another, so we ask tput for that information if available.
* Choose a random sleep value to slightly reduce risk of getting banned by twitch.

glubsy committed:
Suppress backlog clutter

* Overwrite Streamlink's output after every call that doesn't return an exit code of 0.
This should preserve Streamlink's output when a live stream has been detected.
For every other call where a live stream is not active, we safely overwrite the repeating output and keep only the last two lines.
To achieve this we use special non-printable characters to tell the terminal to go back two columns up, clear the line to the beginning, write the date, and clear the rest of the line.
The special characters might be different from one terminal emulator to another, so we ask tput for that information if available.
* Choose a random sleep value to slightly reduce risk of getting banned by twitch.

glubsy commented:
You don't have to merge this if it makes the script too complex for your liking. Just thought it would be useful to know about at least.

mrwnwttk commented:
Very interesting stuff! I'll have to make some changes to make this work on Windows but it'll make its way into the project eventually.

glubsy commented:
Yes I was also thinking about Windows. The only conclusion I came to was to disable this feature altogether on Windows but if you have a better way I'm interested.

glubsy commented:
Looked into it a bit, perhaps Windows uses the same codes (see [Virtual terminal sequences](https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences)), but it might depend on the console being used (I assume either PowerShell or Windows Terminal) and the ENABLE_VIRTUAL_TERMINAL_PROCESSING flag has to be set. No idea.

glubsy committed:
Add fallback values for Windows

* Still no idea how to query terminal capabilities on Windows.
* Fall back to not adding special character sequences on that platform.
* Build sequences that can take parameters (e.g. "cuu").
* Small optimizations.

glubsy committed:
Fix quoted output filename

Avoid writing output filename between quotes and with escaped placeholder variables, i.e. /path/to/"output_filename_[best].ts"

mrwnwttk commented:
Made some adjustments to make everything work on both cmd as well as powershell/windows terminal. Found a way around virtual terminal sequences altogether because having that as a requirement seemed like it was asking a bit much from the end user. Will test with a livestream overnight and merge with my commit on top if it didn't break when I come back to my desk tomorrow.

mrwnwttk commented:
The cmd and powershell fixes were made in https://github.com/mrwnwttk/livestream_scripts/commit/f57da43456d41848194f439a17c57e94d735c159.

glubsy commented:
Nice fixes, thank you!
